src/env.cc

```diff
@@ -1710,11 +1728,23 @@ bool BaseObject::IsRootNode() const {
 }
 
 Local<FunctionTemplate> BaseObject::GetConstructorTemplate(Environment* env) {
-  Local<FunctionTemplate> tmpl = env->base_object_ctor_template();
+  return GetConstructorTemplate(env->isolate());
+}
+
+Local<FunctionTemplate> BaseObject::GetConstructorTemplate(Isolate* isolate) {
+  Local<FunctionTemplate> tmpl =
+      Environment::get_base_object_ctor_template(isolate);
   if (tmpl.IsEmpty()) {
-    tmpl = env->NewFunctionTemplate(nullptr);
-    tmpl->SetClassName(FIXED_ONE_BYTE_STRING(env->isolate(), "BaseObject"));
-    env->set_base_object_ctor_template(tmpl);
+    tmpl = FunctionTemplate::New(isolate,
+                                 nullptr,
+                                 Local<Value>(),
+                                 Local<v8::Signature>(),
+                                 0,
+                                 v8::ConstructorBehavior::kAllow,
+                                 v8::SideEffectType::kHasSideEffect,
+                                 nullptr);
+    tmpl->SetClassName(FIXED_ONE_BYTE_STRING(isolate, "BaseObject"));
+    Environment::set_base_object_ctor_template(isolate, tmpl);
```

Comment:
I am guessing this might be problematic - the builtins aren't really context aware and the code usually assume that there is only one set of these in each environment (which comes from the main context). It may not be enough to simply share the function templates.

Reply:
@joyeecheung

> I am guessing this might be problematic - the builtins aren't really context aware

Why isn't a persistent pointing to the same object enough to share them across contexts if the BaseObject is not bound to a single V8 context?

> the code usually assume that there is only one set of these in each environment (which comes from the main context)

This function does create just one BaseObject ctor template for each environment this is called from, like a singleton class, right?

> It may not be enough to simply share the function templates.

Do you know what else we might need to share here?

Reply:
I was referring to the JSTransferrable constructor template, not sure why I added the comment here. On the other hand,

> Bad news - the kDeserialize method doesn't get called and the 'message' event doesn't get fired on mc.port1 which keeps the event loop alive

I *think* it is caused by not sharing the symbols across worker threads,  it is okay to call Symbol.for in a vm context from the same Environment since they share the same isolate, but if it is called from a worker then the symbol belongs to the symbol table of the worker and won't be recognized by the main instance - you can debug into where env()->messaging_deserialize_symbol() is used to confirm. If that's the case some work needs to be done to ensure that when creating the context for a worker, the symbol from the main context is used instead of being created, maybe an optional parameter passed into NewContext()

Reply:
@joyeecheung for debugging purposes, I tried to replace the symbols with strings because those would not require more work to make make the object cloneable but I'm still seeing the same behavior where the message callback doesn't get called when I send the DOMException object using postMessage(). It probably means that we are not doing something correctly while exposing the JSTransferable ctor to the per_context scripts. Not sure what I'm doing wrong, any clue?

Reply:
@joyeecheung could you PTAL?

Reply:
Is the bug going away? If it is, I assume the templates are shared alright.

Reply:
@joyeecheung unfortunately, nope. :/
The bug still remains, even if we turn the symbol keys into string keys. So that would mean that there's something wrong with the way in which we are sharing the templates.